### PR TITLE
8299915: Remove ArrayAllocatorMallocLimit and associated code

### PR DESCRIPTION
**Pull request by jcking**
Remove abstraction that is a holdover from Solaris. Direct usages of `MmapArrayAllocator` have been switched to normal `malloc`. The justification is that none of the code paths are called from signal handlers, so using `mmap` directly does not make sense and is potentially slower than going through `malloc` which can potentially re-use memory without making any system calls. The remaining usages of `ArrayAllocator` and `MallocArrayAllocator` are equivalent.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299915](https://bugs.openjdk.org/browse/JDK-8299915): Remove ArrayAllocatorMallocLimit and associated code

### Contributors
 * Stefan Karlsson `<stefank@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/11931/head:pull/11931` \
`$ git checkout pull/11931`

Update a local copy of the PR: \
`$ git checkout pull/11931` \
`$ git pull https://git.openjdk.org/jdk.git pull/11931/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11931`

View PR using the GUI difftool: \
`$ git pr show -t 11931`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11931.diff">https://git.openjdk.org/jdk/pull/11931.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/11931#issuecomment-1377906019)

### CODE REVIEW

**Comment by stefank** [comment]
I'm happy to see this flag getting removed. I'm less happy about seeing usages of the array allocators being replaced by macros. I'd rather see an effort towards getting rid of these macros. Could we limit this patch to only remove the ArrayAllocatorMallocLimit flag and ArrayAllocator class, and defer the discussion around what API to use for array allocations?

**Comment by jcking** [comment]
> I'm happy to see this flag getting removed. I'm less happy about seeing usages of the array allocators being replaced by macros. I'd rather see an effort towards getting rid of these macros. Could we limit this patch to only remove the ArrayAllocatorMallocLimit flag and ArrayAllocator class, and defer the discussion around what API to use for array allocations?

`ArrayAllocator` with `ArrayAllocatorMallocLimit` removed is effectively `MallocArrayAllocator`. Are you suggesting leaving `MallocArrayAllocator` and `MmapArrayAllocator` thus update references to `ArrayAllocator` to be `MallocArrayAllocator`?

As far as APIs, the majority of the codebase uses the macros. IMO consistency would be better and having two ways of doing things doesn't help. But if you feel strongly about it, we can punt and just surgically remove the bare minimum, assuming you clarify your expectation (see above).

**Comment by tstuefe** [comment]
Curious, I always thought we do ArrayAllocator - using mmap for larger allocations - to prevent memory retention for libc variants whose allocators are "grabby", i.e. which don't promptly return memory to the OS on free(). E.g. because they only use sbrk (Solaris, AIX), or are just cautious about returning memory (glibc).

Glibc's retention problem is only relevant for fine-grained allocations, so for glibc this is probably fine. This leaves at least AIX as a potential problem. @backwaterred, does the AIX libc still exclusively use sbrk ? Does free'd memory still stick to the process?

(While writing this, I remember that we at SAP even rewrote Arena allocation to use mmap for AIX, because large compile arenas caused lasting RSS increase, so it has definitely been a problem in the past)

**Comment by jcking** [comment]
> > ArrayAllocator with ArrayAllocatorMallocLimit removed is effectively MallocArrayAllocator. Are you suggesting leaving MallocArrayAllocator and MmapArrayAllocator thus update references to ArrayAllocator to be MallocArrayAllocator?
> 
> Yes, that was what I wanted.
> 
> > As far as APIs, the majority of the codebase uses the macros. IMO consistency would be better and having two ways of doing things doesn't help. But if you feel strongly about it, we can punt and just surgically remove the bare minimum, assuming you clarify your expectation (see above).
> 
> I agree about the argument about consistency, so I retract my objection. We can deal with these macros as a separate RFE (if we ever get to it).
> 
> I would like to retain the usage of mmapped memory for ZGC to minimize the risk of any surprises for us. ZGC code also tend to initialize as much as possible in the initialization list, so the extra memset that comes after initialization lists sticks out a bit. Could you create a private `ZGranuleMap::allocate_array` function that uses os::reserve_memory/commmit_memory and change the code to be:
> 
> ```
> inline ZGranuleMap<T>::ZGranuleMap(size_t max_offset) :
>     _size(max_offset >> ZGranuleSizeShift),
>     _map(allocate_array(_size)) {
> ```
> 
> Or if you like, I can provide a patch on top of your branch to do that.

The extra memset is due to `malloc` not handing out zero initialized memory, it looks like `zGranuleMap` is sensitive to that.

Done. Restored mmap as suggested for zGranuleMap.

**Comment by jcking** [comment]
> Curious, I always thought we do ArrayAllocator - using mmap for larger allocations - to prevent memory retention for libc variants whose allocators are "grabby", i.e. which don't promptly return memory to the OS on free(). E.g. because they only use sbrk (Solaris, AIX), or are just cautious about returning memory (glibc).
> 
> Glibc's retention problem is only relevant for fine-grained allocations, so for glibc this is probably fine. This leaves at least AIX as a potential problem. @backwaterred, does the AIX libc malloc() still exclusively use the data segment ? Does free'd memory still stick to the process?
> 
> (While writing this, I remember that we at SAP even rewrote Arena allocation to use mmap for AIX, because large compile arenas caused lasting RSS increase, so it has definitely been a problem in the past)

Correct, glibc always mmap's allocations above a certain limit and always unmaps them on free. I believe most implementations do that, either immediately unmapping or using madvise. AIX still uses sbrk, based on their documentation that I could find. Though AIX does have something called [malloc disclaim](https://www.ibm.com/docs/en/aix/7.2?topic=subsystem-malloc-disclaim).

**Comment by coleenp** [comment]
To follow up on @tstuefe comment - and the one that I tried to say in the bug was that we added this MmapArrayAllocate feature for some G1 marking bits that used so much memory that hit the Solaris _sbrk issue.  Maybe @stefank and @tschatzl remember this issue.  Maybe it's ok for AIX, then removing this code is a good change.  Maybe the G1 usages need a mmap implementation though.

**Comment by jcking** [comment]
> To follow up on @tstuefe comment - and the one that I tried to say in the bug was that we added this MmapArrayAllocate feature for some G1 marking bits that used so much memory that hit the Solaris _sbrk issue. Maybe @stefank and @tschatzl remember this issue. Maybe it's ok for AIX, then removing this code is a good change. Maybe the G1 usages need a mmap implementation though.

The padding.inline.hpp usage seems to have one caller which is called once. The other mmap usage in G1 we can convert to mmap using a similar approach to zGranuleMap if that is preferred. That would then be equivalent behavior, it looks like the G1 code uses the page allocation granularity anyway so maybe keeping it mmap is the better way to go here anyway?

**Review by stefank — CHANGES_REQUESTED** [review]
Thanks for the ZGC update. I have a request to slightly modify the style a bit. Would you mind taking this patch:
https://github.com/stefank/jdk/tree/pr_11931

There's a FIXME about that allocation.inline.hpp file is now empty. Will you remove it in this patch? I'd prefer if you did. Alternatively, remove the FIXME from this patch and then immediately clean this up as a separate patch.

**Comment by tstuefe** [comment]
> > To follow up on @tstuefe comment - and the one that I tried to say in the bug was that we added this MmapArrayAllocate feature for some G1 marking bits that used so much memory that hit the Solaris _sbrk issue. Maybe @stefank and @tschatzl remember this issue. Maybe it's ok for AIX, then removing this code is a good change. Maybe the G1 usages need a mmap implementation though.
> 
> The padding.inline.hpp usage seems to have one caller which is called once. The other mmap usage in G1 we can convert to mmap using a similar approach to zGranuleMap if that is preferred. That would then be equivalent behavior, it looks like the G1 code uses the page allocation granularity anyway so maybe keeping it mmap is the better way to go here anyway?

My uninformed opinion (I'm not the G1 code owner) is that it would be fine to use explicit mmap. I'd love the complexity reduction this patch brings.

**Comment by jcking** [comment]
Applied your patch @stefank

**Comment by jcking** [comment]
Restored `mmap` for G1, though in the future it might make sense to switch to just using malloc.

**Comment by jcking** [comment]
> There's a FIXME about that allocation.inline.hpp file is now empty. Will you remove it in this patch? I'd prefer if you did. Alternatively, remove the FIXME from this patch and then immediately clean this up as a separate patch.

Yeah, separate patch immediately after is probably best. Trying to avoid having to mass update includes and enter merge conflict hell for this patch. After is easier.

**Comment by backwaterred** [comment]
> This leaves at least AIX as a potential problem. @backwaterred, does the AIX libc malloc() still exclusively use the data segment? Does free'd memory still stick to the process?

I agree with @coleenp that AIX seems to still use the data segment and sbrk for dynamic memory allocation. I am not sure about memory sticking to the process.

With these changes I see several of the gtests failing on AIX.

```
[==========] 58 tests from 6 test cases ran. (3718 ms total)
[  PASSED  ] 53 tests.
[  FAILED  ] 5 tests, listed below:
[  FAILED  ] os.safefetchN_negative_vm
[  FAILED  ] os.safefetch32_negative_vm
[  FAILED  ] os.safefetchN_negative_current_null_vm
[  FAILED  ] os.safefetch32_negative_current_null_vm
[  FAILED  ] os.safefetch_negative_at_safepoint_vm
```

I can experiment with the disclaim option as well.

**Comment by iklam** [comment]
@tstuefe @backwaterred I'd like to see this RFE revived. Do we know if anyone is using the `ArrayAllocatorMallocLimit` flag in any production environment today?

It seems unlikely to me, as you'd need to explicitly specify `-XX:+UnlockExperimentalVMOptions` in the command-line.

And, if this option had been useful (for the AIX port, for example), it would have been changed to a non-experimental (with proper `_pd` support) option over the past 10 years.

File: src/hotspot/share/memory/padded.inline.hpp [hunk 1]
```diff
@@ -63,8 +63,11 @@ T** Padded2DArray<T, flags, alignment>::create_unfreeable(uint rows, uint column
   size_t total_size = table_size + rows * row_size + alignment;
 
   // Allocate a chunk of memory large enough to allow alignment of the chunk.
-  void* chunk = MmapArrayAllocator<uint8_t>::allocate(total_size, flags);
+  uint8_t* chunk = NEW_C_HEAP_ARRAY(uint8_t, total_size, flags);
+
   // Clear the allocated memory.
+  memset(chunk, '\0', total_size);
+
```
Comment (tstuefe) [inline src/hotspot/share/memory/padded.inline.hpp:70]: Old code, when doing the malloc path, did not initialize. Why here?

**Comment by tstuefe** [comment]
> @tstuefe @backwaterred I'd like to see this RFE revived. Do we know if anyone is using the `ArrayAllocatorMallocLimit` flag in any production environment today?
> 
> It seems unlikely to me, as you'd need to explicitly specify `-XX:+UnlockExperimentalVMOptions` in the command-line.
> 
> And, if this option had been useful (for the AIX port, for example), it would have been changed to a non-experimental (with proper `_pd` support) option over the past 10 years.

@iklam I'm fine with removing the ArrayAllocator.

**Review by stefank — COMMENTED** [review]
Could we create a simpler version of this PR? One that only removes ArrayAllocator and ArrayAllocatorMallocLimit, but keeps MallocArrayAllocator and MmapArrayAllocator? That way we can figure out later, in a separate RFE, if we want to remove MmapArrayAllocator.

**Pull request closed without merging**